contracts/finance/ConfidentialVestingWallet.sol

```diff
@@ -0,0 +1,125 @@
+        START_TIMESTAMP = startTimestamp_;
+        CONFIDENTIAL_ERC20 = IConfidentialERC20(token_);
+        DURATION = duration_;
+        END_TIMESTAMP = startTimestamp_ + duration_;
+        BENEFICIARY = beneficiary_;
+
+        /// @dev Store this constant variable in the storage.
+        _EUINT64_ZERO = TFHE.asEuint64(0);
+        _amountReleased = _EUINT64_ZERO;
+
+        TFHE.allow(_EUINT64_ZERO, beneficiary_);
+        TFHE.allowThis(_EUINT64_ZERO);
+    }
+
+    /**
+     * @notice  Release the tokens that have already vested.
+     * @dev     Anyone can call this function but the beneficiary receives the tokens.
+     */
+    function release() public virtual {
+        euint64 amount = _releasable();
+        euint64 amountReleased = TFHE.add(_amountReleased, amount);
+        _amountReleased = amountReleased;
+
+        TFHE.allow(amountReleased, BENEFICIARY);
+        TFHE.allowThis(amountReleased);
+        TFHE.allowTransient(amount, address(CONFIDENTIAL_ERC20));
+        CONFIDENTIAL_ERC20.transfer(BENEFICIARY, amount);
+
+        emit ConfidentialERC20Released();
+    }
+
+    /**
+     * @notice                  Return the encrypted amount of total tokens released.
+     * @dev                     It is only reencryptable by the owner.
+     * @return amountReleased   Total amount of tokens released.
+     */
+    function released() public view virtual returns (euint64 amountReleased) {
+        return _amountReleased;
+    }
+
+    /**
+     * @notice                  Calculate the amount of tokens that can be released.
+     * @return releasableAmount Releasable amount.
+     */
+    function _releasable() internal virtual returns (euint64 releasableAmount) {
+        return TFHE.sub(_vestedAmount(uint64(block.timestamp)), released());
+    }
+
+    /**
+     * @notice                  Calculate the amount of tokens that has already vested.
+     * @return vestedAmount     Vested amount.
+     */
+    function _vestedAmount(uint64 timestamp) internal virtual returns (euint64 vestedAmount) {
+        return _vestingSchedule(TFHE.add(CONFIDENTIAL_ERC20.balanceOf(address(this)), released()), timestamp);
+    }
+
+    /**
+     * @notice                  Return the vested amount based on a linear vesting schedule.
+     * @dev                     It must be overriden for non-linear schedules.
+     * @param totalAllocation   Total allocation that is vested.
+     * @param timestamp         Current timestamp.
+     * @return vestedAmount     Vested amount.
+     */
+    function _vestingSchedule(
+        euint64 totalAllocation,
+        uint64 timestamp
+    ) internal virtual returns (euint64 vestedAmount) {
+        if (timestamp < START_TIMESTAMP) {
+            return _EUINT64_ZERO;
+        } else if (timestamp >= END_TIMESTAMP) {
+            return totalAllocation;
+        } else {
+            return TFHE.div(TFHE.mul(totalAllocation, (timestamp - START_TIMESTAMP)), DURATION);
```

Comment:
Warning: there is a critical risk in case of overflow here, you should cast the `totalAllocation` to an euint128 before doing the product this is very important